keps/sig-network/1880-service-subnet/README.md

```diff
@@ -0,0 +1,558 @@
+	// represented as a bit array starting at the base IP of the CIDR in Range, with each bit representing
+	// a single allocated address (the fifth bit on CIDR 10.0.0.0/8 is 10.0.0.4).
+	Data []byte
+}
+```
+
+To avoid leaking resources, they also need to implement to repair loops that run in the `controlplane` instance
+that keep the bitmaps in sync with the current Services allocated IPs.
+
+### Proposed Allocation model
+https://github.com/kubernetes/enhancements/pull/1881#issuecomment-757003362
+
+* define a new Object (i.e. IPRange and NodePortRange) that define the range used for allocation.
+The value of the range is the one passed as parameters in the apiserver flags for backwards compatibility.
+```
+<<[UNRESOLVED range modification]>>
+
+Since the Range objects will be created by the bootstrap logic of the apiserver, the first apiserver to start will define the Range for the whole cluster.
+The Range objects can not be deleted, but can be updated meanwhile the objects allocated referencing it are withing the new range. Per example, in a cluster with a Service Range 10.0.0.0/24, it will be always possible to update the service to a larger subnet 10.0.0.0/16. However, it will not be possible to use a smaller subnet, i.e. 10.0.0.0/28, if there are IPs allocated out if that range, i.e. 10.0.0.244.
+
+<<[/UNRESOLVED]>>
+
+
+```
+
+* define another object (ie. IPAddress and NodePort) that represent the allocated resources, the name of
+the object will be the port or the IP, so we can guarantee the consistency.
+
+* the allocated object must reference the object used to define the range allocation, and validated against it (i.e. IP within range)
+
+```
+<<[UNRESOLVED allocation mode]>>
+
+* and allocated object can have different status: allocated, reserved, free, ...
+
+* on service creation, if the NodePort or ClusterIP field is set:
+1. check that the object doesn't exist in the apiserver and Create a new one (owned by the Service)
+2. check if the object exists and the Status is Free and change the status.
+
+* on service creation, if the NodePort or ClusterIP field is not set, allocate a random one from the defined range in the corresponding Object (IPRange or NodePortRange)
+
+There can be different model to allocate a random resource:
+
+1. Same as today, the apiserver can keep an index with current status and start with a random offset to
+find a free resources.
+
+2. A controller create a pool of free resource Objects and the apiserver picks one of them, this adds 
+more complexity to the model, but will improve the time to allocate a new free resources since is just
+popping an object from a list.
+
+<<[/UNRESOLVED]>>
+```
+
+## Design Details
+
+
+### Cluster IPs
+
+#### IP Ranges
+
+Define two new objects, a new IPRange Object that defines the range used to allocate Services Cluster IPs
+
+```go
+// IPRange defines a range of IPs using CIDR format (192.168.0.0/24 or 2001:db2::0/64).
+type IPRange struct {
```

Comment:
Need to talk about overlaps - are they allowed and what do they mean?  As my above notes, I think overlap is OK and just gets merged by anyone using this API

Reply:
we  can validate overlaps if we use status on the IPRange

1. Create IPRange (simply validation= -> status "notReady"
2. Controller sees the new IPRange, performs validation against all the IPRanges
3. No overlapping, set status to Ready, it can start allocating
4. Overlapping, keep status as notReady ot invalid, and a condition message